### PR TITLE
[PERF]MiniMax-M2 gate kernel

### PR DESCRIPTION
**Pull request by jeejeelee**
## Purpose

The MoE gate of MiniMax-M2 needs to perform GEMM in FP32 and requires converting the input to FP32, see [router_logits](https://github.com/vllm-project/vllm/blob/v0.18.1rc0/vllm/model_executor/models/minimax_m2.py#L132). Therefore, in low-concurrency scenarios, up to 3 kernels may be launched in the worst case:
-  the first one is BF16-to-FP32 conversion
- and the second and third are the two kernels of `split-K` GEMM. 

img_v3_02108_87c02d83-ccac-4465-8f0e-4b28a96fc7eg

This PR implements a fused kernel referencing the `dsv3_router_gemm` implementation, which yields a significant performance improvement under low-concurrency conditions.

img_v3_02108_ad60c4ed-dc70-4432-9b7b-9730440ced6g

## Benchmark Config

**Hardware**
4x NVIDIA GB200 (189471 MiB each)

---

### Branch Info

| | Baseline | Feature |
|---|---|---|
| **Branch** | `main` | `m2-gate-function` |
| **Commit** | `b2bc736b1` | `fe8426d3c` |
| **vLLM version** | `0.1.dev15327+gb2bc736b1` | `0.1.dev15334+gfe8426d3c` |

---

### GSM8K (5-shot)

| Metric | main | feature | Δ | % Change |
|---|---:|---:|---:|---:|
| flexible-extract | 0.9401 | 0.9318 | -0.0083 | -0.9% |
| strict-match | 0.9378 | 0.9280 | -0.0098 | -1.0% |

### AIME25 (epoch_num=4)

| Metric | main | feature | Δ | % Change |
|---|---:|---:|---:|---:|
| exact_match | 0.8417 | 0.8333 | -0.0084 | -1.0% |

### GPQA Diamond (epoch_num=4)

| Metric | main | feature | Δ | % Change |
|---|---:|---:|---:|---:|
| exact_match | 0.8182 | 0.8384 | +0.0202 | **+2.5%** |

**Key observations:**

- **No accuracy regression detected.** 

---

## Performance Comparison

**Server script**
- Used the internal bench tool, the startup script is similar:

```bash
      export VLLM_FLASHINFER_MOE_BACKEND=latency
      vllm serve nvidia/MiniMax-M2.5-NVFP4
      -tp 4
      --performance-mode interactivity
      --trust-remote-code
      --max-num-seq 64
      --kv-cache-dtype fp8
      --compilation-config '{"mode":3,"pass_config":{"fuse_norm_quant":true,"fuse_act_quant":true,"fuse_gemm_comms":true}}'
```

**Workload**
SPEED-Bench throughput_16k / low_entropy, input ≤10240 tokens, output 1536 tokens, concurrency sweep 1–64

### Output Token Throughput (tok/s) — higher is better

| Concurrency | main | feature | Δ tok/s | % Change |
|---|---:|---:|---:|---:|
| 1 | 147.56 | 157.04 | +9.48 | **+6.4%** |
| 2 | 234.32 | 309.81 | +75.49 | **+32.2%** |
| 4 | 510.02 | 553.13 | +43.11 | **+8.5%** |
| 8 | 959.10 | 1031.65 | +72.55 | **+7.6%** |
| 16 | 1634.49 | 1731.21 | +96.72 | **+5.9%** |
| 32 | 2381.54 | 2720.66 | +339.12 | **+14.2%** |
| 64 | 3965.02 | 3958.64 | -6.38 | -0.2% |

### Total Token Throughput (tok/s) — higher is better

| Concurrency | main | feature | Δ tok/s | % Change |
|---|---:|---:|---:|---:|
| 1 | 1192.55 | 1203.97 | +11.42 | **+1.0%** |
| 2 | 1796.43 | 2439.87 | +643.44 | **+35.8%** |
| 4 | 3945.96 | 4404.30 | +458.34 | **+11.6%** |
| 8 | 7420.80 | 8047.16 | +626.36 | **+8.4%** |
| 16 | 12633.50 | 13443.59 | +810.09 | **+6.4%** |
| 32 | 18480.13 | 21193.10 | +2712.97 | **+14.7%** |
| 64 | 30944.06 | 30966.74 | +22.68 | +0.1% |

### Mean TTFT (ms) — lower is better

| Concurrency | main | feature | Δ ms | % Change |
|---|---:|---:|---:|---:|
| 1 | 236.38 | 212.75 | -23.63 | **-10.0%** |
| 2 | 340.99 | 242.06 | -98.93 | **-29.0%** |
| 4 | 346.41 | 319.68 | -26.73 | **-7.7%** |
| 8 | 454.57 | 449.39 | -5.18 | **-1.1%** |
| 16 | 467.95 | 514.96 | +47.01 | +10.0% ⚠️ |
| 32 | 597.40 | 572.90 | -24.50 | **-4.1%** |
| 64 | 613.47 | 608.84 | -4.63 | -0.8% |

### Mean TPOT (ms) — lower is better

| Concurrency | main | feature | Δ ms | % Change |
|---|---:|---:|---:|---:|
| 1 | 6.62 | 6.23 | -0.39 | **-5.9%** |
| 2 | 8.32 | 6.22 | -2.10 | **-25.2%** |
| 4 | 7.53 | 6.73 | -0.80 | **-10.6%** |
| 8 | 7.96 | 7.32 | -0.64 | **-8.0%** |
| 16 | 9.40 | 8.79 | -0.61 | **-6.5%** |
| 32 | 12.89 | 11.20 | -1.69 | **-13.1%** |
| 64 | 15.45 | 15.47 | +0.02 | +0.1% |

### Request Throughput (req/s) — higher is better

| Concurrency | main | feature | % Change |
|---|---:|---:|---:|
| 1 | 0.10 | 0.10 | 0% |
| 2 | 0.15 | 0.21 | **+40.0%** |
| 4 | 0.34 | 0.38 | **+11.8%** |
| 8 | 0.63 | 0.69 | **+9.5%** |
| 16 | 1.07 | 1.14 | **+6.5%** |
| 32 | 1.57 | 1.80 | **+14.6%** |
| 64 | 2.63 | 2.64 | +0.4% |

---

### Summary

`m2-gate-function` demonstrates consistent improvements over `main` across low-to-mid concurrency workloads:

| Metric | Low concurrency (1–4) | Mid concurrency (8–32) | High concurrency (64) |
|---|---|---|---|
| Output tok/s | +6% ~ +32% | +6% ~ +14% | <1% |
| TTFT | -8% ~ -29% | -1% ~ -4% (⚠️ +10% at c=16) | <1% |
| TPOT | -6% ~ -25% | -7% ~ -13% | <1% |

**Key observations:**

- **Concurrency=2** shows the largest gains: +32% output throughput, -29% TTFT, -25% TPOT — the feature branch most benefits low-concurrency interactive serving.
- **Concurrency=32** shows significant throughput improvement (+14.2%) with better TPOT (-13.1%), suggesting improved kernel efficiency under moderate load.
- **Concurrency=64**: All metrics differ by <1% between branches. The operator changes do not affect the high-concurrency saturation path; these differences are considered statistical noise.
- **TTFT regression at concurrency=16** (+10%, +47ms): the only anomaly; worth investigating further.
- No failures across all 7 concurrency levels on both branches.

### CODE REVIEW

File: vllm/model_executor/models/minimax_m2.py [hunk 1]
```diff
@@ -127,9 +127,11 @@ def ebias_weight_loader(param: nn.Parameter, loaded_weight: torch.Tensor) -> Non
     def forward(self, hidden_states: torch.Tensor) -> torch.Tensor:
         num_tokens, hidden_dim = hidden_states.shape
         hidden_states = hidden_states.view(-1, hidden_dim)
-
+        if num_tokens <= 32:
+            pass
```
[the remark this answers is not in the document]
Comment (jeejeelee) [reply vllm/model_executor/models/minimax_m2.py]: Removed

**Comment by jeejeelee** [comment]
@claude review

**Comment by qianlihuang** [comment]
Hi @jeejeelee nice work on this PR. I noticed it's been open for a while. Any remaining blockers? Happy to help

**Comment by jeejeelee** [comment]
I wiil restart this PR ASAP

File: cmake/utils.cmake [hunk 2]
```diff
@@ -476,6 +476,16 @@ function(cuda_archs_loose_intersection OUT_CUDA_ARCHS SRC_CUDA_ARCHS TGT_CUDA_AR
   set(${OUT_CUDA_ARCHS} ${_CUDA_ARCHS} PARENT_SCOPE)
 endfunction()
 
+
+function(cuda_archs_sm90plus OUT_CUDA_ARCHS TGT_CUDA_ARCHS)
+  if(${CMAKE_CUDA_COMPILER_VERSION} VERSION_GREATER_EQUAL 13.0)
+    cuda_archs_loose_intersection(_archs "9.0a;10.0f;11.0f" "${TGT_CUDA_ARCHS}")
+  else()
+    cuda_archs_loose_intersection(_archs "9.0a;10.0a;10.1a;10.3a" "${TGT_CUDA_ARCHS}")
```
Comment (Harry-Chen) [inline cmake/utils.cmake:484]:
```suggestion
    cuda_archs_loose_intersection(_archs "9.0a;10.0f;11.0f;12.0f" "${TGT_CUDA_ARCHS}")
  else()
    cuda_archs_loose_intersection(_archs "9.0a;10.0a;10.1a;10.3a;12.0a;12.1a" "${TGT_CUDA_ARCHS}")
```

Maybe let's include SM120 family support?

Comment (jeejeelee) [reply cmake/utils.cmake:484]: done

**Review by ywang96 — APPROVED** [review]
